subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -232,4 +274,43 @@ public Configuration getCompileDependencies() {
             return convention.getProject().getConfigurations().getByName(JavaPlugin.COMPILE_CONFIGURATION_NAME);
         }
     }
+
+    /**
+     * A custom artifact type which allows the getFile call to be done lazily only when the
+     * artifact is actually needed.
+     */
+    private abstract static class IntermediateJavaArtifact extends AbstractPublishArtifact {
```

Comment:
I think we should address this as part of this story instead of adding more complexity in the Java plugin. People will stumble over this code and think it is the way to go, adding more unwanted subclasses of `PublishArtifact`